README.md
# What is Sam Drucker?

SamDrucker is a collection of small components which create a centralized
list of all packages on all hosts.

Each component is designed to be:

* small
* simple
* easily written
* flexiable
* few, if any, dependencies

## The components

* client - collects the list of packages and posts it to the server
* server - accepts the list of packages and adds it to the database
* database - tables and stored procedures for the catalog of packages

Each component can be written in whatever languages you want. Collect
the packages in any manner you want. This can be done remotely on the host
or centrally on a management tool, such as Ansible.

Pick whatever languages you want.

## Other ideas

It was mentioned elsewhere that Ansible or Spacewalk can help here. I want
something completely independent. These tools are great at collecting information.

I didn't want to use a large number of dependencies or huge packages.

## Current status

A sample client and web service have been created as proof of concept.
They are deployed in my home network, active on about 90 hosts.

A database schema has been created for PostgreSQL, my database of personal
choice, but this should work on any other database. Patches welcome here.

## Next major step

At present, queries of the database are possible only via command line tools.

I'd like to create a simple web interface.

Anyone is weclome to help write this stuff.


## Wiki

You can read more in [the wiki](https://github.com/dlangille/SamDrucker/wiki).


CHANGELOG.md
# Changelog

## host table

Any changes to the `host` table with automatically update the `date_updated`
field. See `update_date_updated` in `PostgreSQL/database-postgresql.sql`.

## 0.2.4

With this release, you can set a host to be not enabled. Perhaps you have a
host which is no longer in use but you want to keep the data.

When upgrading to this release, be sure to run these scripts:

* `PostgreSQL/updates-2020.08.28.sql`
* `PostgreSQL/procedures.sql`

If you want to disable any hosts, this SQL will work:

```
update host set enabled = false where name in ('foo', 'bar');
```

New versions of the query functions have been created. These take a boolean
parameter as their last arguement. This allows you to query hosts which are
no longer enabled (false). The existing functions will now only query
enabled hosts. You can retain existing behaviour by leaving all hosts
enabled.


clients/samdrucker.sh
#!/bin/sh

# we want to construct JSON which looks like this:
# {
#   "name": "test.example.org",
#   "os": "FreeBSD",
#   "version": "11.3-RELEASE-p4",
#   "repo": "http://pkg.freebsd.org/FreeBSD:113:amd64/latest/",
#   "packages": [
#       "apr-1.6.5.1.6.1_1",
#       "bacula9-client-9.4.3",
#       "bash-5.0.7",
#       "rsync-3.1.3_1",
#       "serf-1.3.9_3",
#       "sqlite3-3.29.0_1"
#    ]
# }
#
# I am using the FreeBSD textproc/jo:
#
# $ jo -p name=`hostname` os=`uname` version=`uname -r` repo=`pkg -vv | grep url | cut -f2 -d \"`
# {
#    "name": "samdrucker.int.unixathome.org",
#    "os": "FreeBSD",
#    "version": "12.0-RELEASE-p10",
#    "repo": "pkg+http://fedex.unixathome.org/packages/120amd64-default-master-list/"
# }
#

if [ -r /usr/local/etc/samdrucker/samdrucker.conf ]; then
  . /usr/local/etc/samdrucker/samdrucker.conf
fi

CURL="/usr/local/bin/curl"
CUT="/usr/bin/cut"
GREP="/usr/bin/grep"
JO="/usr/local/bin/jo"
PKG="/usr/sbin/pkg"

# get list of packages on this host:

pkg_args=""
PKGS=`$PKG info -q`
for pkg in $PKGS
do
  pkg_args="$pkg_args packages[]=$pkg"
done

hostname=`hostname`
uname=`uname`
version=`uname -r`
repo=`/usr/sbin/pkg -vv | $GREP  url | $CUT -f2 -d \"`

# we save this to a file to avoid potential command line arguement overflow
payload=$(mktemp /tmp/SamDrucker.payload.XXXXXX)
$JO -p name=$hostname os=$uname version=$version repo=$repo $pkg_args > $payload

$CURL $CURL_OPTIONS --data-urlencode ${SAMDRUCKER_ARG}@${payload} $SAMDRUCKER_URL

# remove the temp file
rm $payload


servers/php/samdrucker.php
<?php

define('CONF_FILE', '/usr/local/etc/samdrucker/samdruckerserver.conf');

if (file_exists(CONF_FILE)) {
  require_once(CONF_FILE);
} else {
  syslog(LOG_ERR, CONF_FILE . ' does not exist');
  die(CONF_FILE . ' does not exist');
  exit;
}

json_decode($_REQUEST['packages']);
if (json_last_error() == JSON_ERROR_SYNTAX) {
  # does this need some sanitizing?
  syslog(LOG_ERR, 'invalid json: ' . $_REQUEST['packages']);
  echo 'invalid JSON';
  exit;
}

$dsn = "pgsql:host=$host;port=$port;dbname=$db;user=$user;password=$password;sslmode=require";

try {
  $dbh = new PDO($dsn);

  $dbh->beginTransaction();
  $sql = 'SELECT HostAddPackages(:packages, :client_ip)';

  $stmt = $dbh->prepare($sql);
  # does this need some sanitizing?
  $stmt->bindValue(':packages',  $_REQUEST['packages']);
  $stmt->bindValue(':client_ip', $_SERVER['REMOTE_ADDR']);

  try {
    $stmt->execute();
  }  catch (PDOException $e){
    echo $e->getMessage();
  }

  $dbh->commit();

} catch (PDOException $e){
 // report error message
 syslog(LOG_ERR, $e->getMessage());
}


clients/samdrucker-py.sh
#!/bin/sh

# we want to construct JSON which looks like this:
# {
#   "name": "test.example.org",
#   "os": "FreeBSD",
#   "version": "11.3-RELEASE-p4",
#   "repo": "http://pkg.freebsd.org/FreeBSD:113:amd64/latest/",
#   "packages": [
#       "apr-1.6.5.1.6.1_1",
#       "bacula9-client-9.4.3",
#       "bash-5.0.7",
#       "rsync-3.1.3_1",
#       "serf-1.3.9_3",
#       "sqlite3-3.29.0_1"
#    ]
# }
#
# I am using the FreeBSD textproc/jo:
#
# $ jo -p name=`hostname` os=`uname` version=`uname -r` repo=`pkg -vv | grep url | cut -f2 -d \"`
# {
#    "name": "samdrucker.int.unixathome.org",
#    "os": "FreeBSD",
#    "version": "12.0-RELEASE-p10",
#    "repo": "pkg+http://fedex.unixathome.org/packages/120amd64-default-master-list/"
# }
#

if [ -r /usr/local/etc/samdrucker/samdrucker.conf ]; then
  . /usr/local/etc/samdrucker/samdrucker.conf
fi

# get list of packages on this host:

pkg_args=""
PKGS=`/usr/sbin/pkg info -q`
for pkg in $PKGS
do
  pkg_args="$pkg_args packages[]=$pkg"
done

#echo $pkg_args

hostname=`hostname`
uname=`uname`
version=`uname -r`
repo=`pkg -vv | grep url | cut -f2 -d \"`

payload=`/usr/local/bin/jo -p name=$hostname os=$uname version=$version repo=$repo $pkg_args`

#echo "$payload"

# I tried but failed with this attempt
#encoded=`echo "$payload" | php -R 'echo rawurlencode(fgets(STDIN));'`

# So I went with python
#encoded=`/usr/local/bin/python -c "import urllib; print urllib.quote('''$payload''')"`

#echo "$encoded"

curl --output /dev/null  -d "$SAMDRUCKER_ARG=$payload" -H "Content-Type: application/x-www-form-urlencoded" -X POST $SAMDRUCKER_URL
